### ISSUE TITLE
Joins return incorrect results if performed after index map

### ISSUE THREAD

**Issue by velezsarain** [dask/dask#6265]
**What happened**: Joining dask objects return incorrect results if the join operation was performed after the index of the dask object was modified by a map operation beforehand.  Resetting then setting the index again fixes this issue, but documentation does not seem to mention if this behaviour is intended or not

**What you expected to happen**: The join should work as if it was performed in pandas, taking into account the new index values.

**Minimal Complete Verifiable Example**:

```python
import dask
import numpy as np
import pandas as pd

original = dask.datasets.timeseries()
another_index = pd.date_range(
    start=np.datetime64('1999-12-31 00:00:00'), end=np.datetime64('2000-01-29 23:59:59'), freq='1s')
print(len(original.index.compute().join(another_index, how='inner')))

original.index = original.index.map(
    lambda x: x - np.timedelta64(1, 'D'), meta=('timestamp', 'datetime64[ns]'))
another_dataframe = pd.DataFrame(index=another_index)
print(len(original.index.compute().join(another_index, how='inner')))

first_join = original.join(another_dataframe, how='inner')
print(len(first_join.index.compute()))

original = original.reset_index()
original = original.set_index('timestamp', npartitions='auto')
second_join = original.join(another_dataframe, how='inner')
print(len(second_join.index.compute()))
```

Output :
2505600 --- Old and new index have overlapping elements
2592000 --- After mapping indexes are equivalent
0 --- First dask join fails, does not take into account the old or new index
2592000 --- Correct result only after resetting and setting the index again

**Environment**:

- Dask version: 2.15.0
- Python version: 3.6.9
- Operating System: Ubuntu 18.04.4 LTS
- Install method (conda, pip, source): pip

**Comment by jsignell** [comment]
Hi @velezsarain thanks for opening this issue! This is a real issue although it's not clear to me how to address it. I guess we need to be sure to do the index assignment before we do the any joins. I have been looking at the task graphs (`.visualize()`) of `original` and `first_join` to try to get some insight into what is going on. I will be able to take a deeper look tomorrow.

**Comment by TomAugspurger** [comment]
@velezsarain are your index partitions valid after assigning the index?

If I clear the divisions then the numbers seem correct

```
In [24]: original = original.clear_divisions()

In [25]: another_dataframe = pd.DataFrame(index=another_index)
    ...: print(len(original.index.compute().join(another_index, how='inner')))
    ...:
    ...: first_join = original.join(another_dataframe, how='inner')
    ...: print(len(first_join.index.compute()))
    ...:
    ...: original = original.reset_index()
    ...: original = original.set_index('timestamp', npartitions='auto')
    ...: second_join = original.join(another_dataframe, how='inner')
    ...: print(len(second_join.index.compute()))
2592000
2592000
2592000
```

I'd *strongly* recommend against setting the index like that. Better to use `set_index`, which I think would have handled this for you.

```python
In [30]: original = original.set_index(original.index.map(
    ...:     lambda x: x - np.timedelta64(1, 'D'), meta=('timestamp', 'datetime64[ns]')))
    ...:
    ...:

In [31]: original.divisions
Out[31]:
(Timestamp('1999-12-31 00:00:00', freq='S'),
 Timestamp('2000-01-01 00:00:00', freq='S'),
 Timestamp('2000-01-02 00:00:00', freq='S'),
...
```

It *might* be worth overriding `_Frame.index.setter` to not allow setting via `.`. Do you have thoughts on that @jsignell?

**Comment by TomAugspurger** [comment]
Looking briefly at `index.setter`

```python
    @index.setter
    def index(self, value):
        self.divisions = value.divisions
        result = map_partitions(
            methods.assign_index, self, value, enforce_metadata=False
        )
        self.dask = result.dask
        self._name = result._name
        self._meta = result._meta
```

the issue points to the *real* problem, that the divisions in `index.map(...)` are incorrect. If they were correct, then things would be fine. In `map_partitions` we have a warning

```
        Note that the index and divisions are assumed to remain unchanged.
```

We might consider overriding the docstring for `Index.map` to include a similar warning (right now Index.map is just Series.map).

**Comment by velezsarain** [comment]
@TomAugspurger You are correct, I assumed that this issue wasn't a bug, but I found nothing when I searched the documentation for the index. In pandas setting the index in this manner does not lead to errors, leading to my mistake.

**Comment by TomAugspurger** [comment]
I think the best short-term option is to define `Index.map` with a docstring that contains a similar warning to `.map_partitions`, that the divisions are assumed to be unchanged, and should be cleared or set manually if they are (or, I guess another option is to apply the mapping to the scalar division values? I'm not sure how I feel about that.)

**Comment by jsignell** [comment]
Thanks @TomAugspurger! I would be fine with not allowing dot assignment for indexes if it is not recommended and can lead to bad results. I'd rather throw an error and tell people to use `set_index` instead. I think a warning in `index.map` would also be a good addition.

**Comment by velezsarain** [comment]
Unfortunately while fixing this issue on my project I ran into another issue with the join operation. I'm currently trying to reproduce to see if it is related to index mapping but the gist of it is that an inner join is duplicating one of my dataframes, returning me more items that I had to begin with.

**Comment by velezsarain** [comment]
I've managed to reproduce, but I figured out it is not related to index mapping, so I will be opening another issue for this bug.

**Comment by jsignell** [comment]
I am thinking about this again and if you call `map` on an index isn't it extremely likely that you are changing the divisions? Maybe the better approach is to always run `.clear_divisions()` on the `.map` method.

**Comment by TomAugspurger** [comment]
> I am thinking about this again and if you call map on an index isn't it extremely likely that you are changing the divisions? 

I would think so. *If* we wanted to be clever, we could call / lookup the mapping on the divisions specifically, and assign the new divisions to the result of that. Something like

```python
new_divisions = pd.Series(self.divisions).map(func)
```

but I'd be hesitant to do that since it's assuming the mapping is monotonic.

**Comment by jsignell** [comment]
Yeah plus you can map with an iterable rather than a function which would make no sense in that case. I've updated the PR: #6285 to just clear_divisions every time `index.map` is called.
